pkg/relayer/session/claim.go

```diff
@@ -15,132 +15,167 @@ import (
 		ctx, sessionsWithOpenClaimWindowObs,
-		rs.newMapClaimSessionFn(failedCreateClaimSessionsPublishCh),
+		rs.newMapClaimSessionsFn(failedCreateClaimSessionsPublishCh),
 	)
 
 	// TODO_TECHDEBT: pass failed create claim sessions to some retry mechanism.
 	_ = failedCreateClaimSessionsObs
 	logging.LogErrors(ctx, filter.EitherError(ctx, eitherClaimedSessionsObs))
 
-	// Map eitherClaimedSessions to a new observable of relayer.SessionTree which
-	// is notified when the corresponding claim creation succeeded.
+	// Map eitherClaimedSessions to a new observable of []relayer.SessionTree
+	// which is notified when the corresponding claims creation succeeded.
 	return filter.EitherSuccess(ctx, eitherClaimedSessionsObs)
 }
 
-// mapWaitForEarliestCreateClaimHeight is intended to be used as a MapFn. It
-// calculates and waits for the earliest block height, allowed by the protocol,
-// at which a claim can be created for the given session, then emits the session
-// **at that moment**.
-func (rs *relayerSessionsManager) mapWaitForEarliestCreateClaimHeight(
-	ctx context.Context,
-	session relayer.SessionTree,
-) (_ relayer.SessionTree, skip bool) {
-	rs.waitForEarliestCreateClaimHeight(
-		ctx, session.GetSessionHeader().GetSessionEndBlockHeight(),
-	)
-	return session, false
+// mapWaitForEarliestCreateClaimsHeight is intended to be used as a MapFn.
+// It calculates and waits for the earliest block height, allowed by the protocol,
+// at which claims can be created for the given session number, then emits the
+// session **at that moment**.
+func (rs *relayerSessionsManager) mapWaitForEarliestCreateClaimsHeight(
+	failedCreateClaimsSessionsPublishCh chan<- []relayer.SessionTree,
+) channel.MapFn[[]relayer.SessionTree, []relayer.SessionTree] {
+	return func(
+		ctx context.Context,
+		sessionTrees []relayer.SessionTree,
+	) (_ []relayer.SessionTree, skip bool) {
+		return rs.waitForEarliestCreateClaimsHeight(
+			ctx, sessionTrees, failedCreateClaimsSessionsPublishCh,
+		), false
+	}
 }
 
-// waitForEarliestCreateClaimHeight calculates and waits for (blocking until) the
-// earliest block height, allowed by the protocol, at which a claim can be created
+// waitForEarliestCreateClaimsHeight calculates and waits for (blocking until) the
+// earliest block height, allowed by the protocol, at which claims can be created
 // for a session with the given sessionEndHeight. It is calculated relative to
 // sessionEndHeight using on-chain governance parameters and randomized input.
 // It IS A BLOCKING function.
-func (rs *relayerSessionsManager) waitForEarliestCreateClaimHeight(
+func (rs *relayerSessionsManager) waitForEarliestCreateClaimsHeight(
 	ctx context.Context,
-	sessionEndHeight int64,
-) {
-	logger := polylog.Ctx(ctx)
+	sessionTrees []relayer.SessionTree,
+	failSubmitProofsSessionsCh chan<- []relayer.SessionTree,
+) []relayer.SessionTree {
+	sessionEndHeight := sessionTrees[0].GetSessionHeader().GetSessionEndBlockHeight()
```

Comment:
Are we making an assumption that all session trees share the same session number? If so, what makes us certain of that?

Reply:
Updated the logic on `mapBlockToSessionsToClaim` (`pkg/relayer/session/session.go`) to emit sessionTree groups that have the same `SessionEndBlockHeight`.